packages/protocol/contracts/verifiers/compose/ComposeVerifier.sol

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+import "../../common/EssentialContract.sol";
+import "../IVerifier.sol";
+
+/// @title ComposeVerifier
+/// @notice This contract is an abstract verifier that composes multiple sub-verifiers to validate
+/// proofs.
+/// It ensures that a set of sub-proofs are verified by their respective verifiers before
+/// considering the overall proof as valid.
+/// @custom:security-contact security@taiko.xyz
+abstract contract ComposeVerifier is EssentialContract, IVerifier {
+    struct SubProof {
+        address verifier;
+        bytes proof;
+    }
+
+    error INSUFFICIENT_PROOF();
+
+    event SubProofError(address indexed verifier, bytes returnData);
+
+    /// @notice Initializes the contract.
+    /// @param _owner The owner of this contract. msg.sender will be used if this value is zero.
+    /// @param _rollupAddressManager The address of the {AddressManager} contract.
+    function init(address _owner, address _rollupAddressManager) external initializer {
+        __Essential_init(_owner, _rollupAddressManager);
+    }
+
+    /// @notice Verifies one or more sub-proofs.
+    /// @param _ctx The context of the proof verification.
+    /// @param _tran The transition to verify.
+    /// @param _proof The proof to verify.
+    function verifyProof(
+        Context calldata _ctx,
+        TaikoData.Transition calldata _tran,
+        TaikoData.TierProof calldata _proof
+    )
+        external
+    {
+        (address[] memory verifiers, uint256 threshold) = getSubVerifiersAndThreshold();
+
+        for (uint256 i; i < verifiers.length; ++i) {
+            // Store the value 1 in the temporary storage slot using inline assembly
```

Comment:
Should not we have a safeguard to not possibly reuse verifiers ? 🤔
 
Something like:
```
address prevVerifier;
for (uint256 i; i < verifiers.length; ++i) {
    if (uint160(prevVerifier) >= uint160(verifiers[i])) {
                    revert L1_INVALID_VERIFIER();
    }
    ...
    prevVerifier = verifiers[i];
}
```

Reply:
The verifier is coded by us and is not dynamically configured by proposers/provers.